src/io/read_tree_genesis_hdf5.c

```diff
@@ -0,0 +1,1142 @@
+
+
+    //First populate the NextProg
+    for(int64_t i=0;i<nhalos;i++) {
+        int32_t desc = local_halos[i].Descendant;
+        if(desc == -1) continue;
+
+        XRETURN(desc >=0 && desc < nhalos, -1,
+                "Error: for halonum = %"PRId64" at snapshot = %d with ID = %lld "
+                "(forestnr = %"PRId64") the descendant = %d must be located within [0, %"PRId64")\n",
+                i, local_halos[i].SnapNum, local_halos[i].MostBoundID, forestnr, desc, nhalos);
+
+        int32_t first_prog_of_desc_halo = local_halos[desc].FirstProgenitor;
+        if(first_prog_of_desc_halo == -1) {
+            //THIS can not happen. FirstProg should have been assigned correctly already
+            fprintf(stderr,"Error: FirstProgenitor can not be -1\n");
+            fprintf(stderr,"Forestnr = %"PRId64" descendant halo number = %d (at descendant snap = %d) \n", forestnr, desc, local_halos[desc].SnapNum);
+            fprintf(stderr,"ID of this (i=%"PRId64") halo = %lld, ID of descendant halo = %lld\n",
+                    i, local_halos[i].MostBoundID, local_halos[desc].MostBoundID);
+            return -1;
+        }
+
+        //if the first progenitor is this current halo, then nothing to do here
+        if(first_prog_of_desc_halo == i) continue;
+
+        //So the current halo is not the first progenitor - so we need to assign this current halo
+        int32_t next_prog = first_prog_of_desc_halo;
+        while(local_halos[next_prog].NextProgenitor != -1) {
+            next_prog = local_halos[next_prog].NextProgenitor;
+            if((next_prog != -1) && (next_prog < 0 || next_prog >= nhalos)) {
+                fprintf(stderr,"Error: In function %s> next_prog = %d should be either -1 or must be within [0, %"PRId64")\n"
+                        "forestnr = %"PRId64" halonum = %"PRId64" haloid = %lld at snap = %d "
+                        "descendant = %d at snap = %d descID = %lld first_prog_of_desc_halo = %d \n",
+                        __FUNCTION__, next_prog, nhalos, forestnr, i, local_halos[i].MostBoundID, local_halos[i].SnapNum,
+                        desc, local_halos[desc].SnapNum, local_halos[desc].MostBoundID, first_prog_of_desc_halo);
+                return -1;
+            }
+        }
+        if(next_prog < 0 || next_prog >= nhalos || local_halos[next_prog].NextProgenitor != -1) {
+            fprintf(stderr,"Error: In function %s> Bug in code logic. next_prog = %d must be within [0, %"PRId64") "
+                    "*AND* local_halos[next_prog].NextProgenitor = %d should be '-1'",
+                    __FUNCTION__, next_prog, nhalos, local_halos[next_prog].NextProgenitor);
+            return -1;
+        }
+
+        local_halos[next_prog].NextProgenitor = i;
+    }
+
+    //Now populate the NextHaloinFOFGroup
+    for(int64_t i=0;i<nhalos;i++) {
+        int32_t fofhalo = local_halos[i].FirstHaloInFOFgroup;
+        if(fofhalo == -1 || fofhalo < 0 || fofhalo >= nhalos) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Redundant condition: If 'fofhalo == -1', the comparison 'fofhalo < 0' is always true.](https://app.codacy.com/manual/jacobseiler/sage-model/pullRequest?prid=4620074)